### PR TITLE
WIP: Filter labels for matching series by name

### PR DESCRIPTION
**Pull request by boyskila**
Fixes: https://github.com/prometheus/prometheus/issues/6178

I'm going to label this PR as WIP since IMO the filter can be extended to handle more scenarios like those described [here](https://docs.google.com/document/d/1sbR7WLWrh1f6ptxX3-92sNM4aP1skc5R44LxnikxGVY/edit) and for this reason, I'm skipping tests and docs examples for now.
@krasi-georgiev @juliusv @brancz What is your opinion?

UPDATE [@krasi-georgiev ]
Here is a simple doc with few examples
https://docs.google.com/document/d/1sbR7WLWrh1f6ptxX3-92sNM4aP1skc5R44LxnikxGVY/edit

### CODE REVIEW

File: web/api/v1/api.go [hunk 1]
```diff
@@ -477,13 +477,93 @@ func (api *API) labelNames(r *http.Request) apiFuncResult {
...
 
+var (
+	minTime = time.Unix(math.MinInt64/1000+62135596801, 0).UTC()
+	maxTime = time.Unix(math.MaxInt64/1000-62135596801, 999999999).UTC()
+
+	minTimeFormatted = minTime.Format(time.RFC3339Nano)
+	maxTimeFormatted = maxTime.Format(time.RFC3339Nano)
+)
+
+func parseTimeParam(r *http.Request, paramName string, defaultValue time.Time) (time.Time, error) {
```
Comment (krasi-georgiev) [inline web/api/v1/api.go]: to speed up the reviews such changes that can be indipendent from the main PR could be opened as a separate PR

Comment (boyskila) [reply web/api/v1/api.go]: Ok, understand. What do you suggest - to revert this back and to open another PR?

Comment (krasi-georgiev) [reply web/api/v1/api.go]: I think it might be fine, the PR is small enough as is.

Comment (krasi-georgiev) [reply web/api/v1/api.go]: why not return timestamps here to avoid the duplicate call to `timestamp.FromTime`

File: web/api/v1/api.go [hunk 2]
```diff
@@ -477,13 +477,93 @@ func (api *API) labelNames(r *http.Request) apiFuncResult {
...
+		}
+
+		q, err := api.Queryable.Querier(ctx, timestamp.FromTime(start), timestamp.FromTime(end))
+		defer q.Close()
+
+		if err != nil {
+			return apiFuncResult{nil, &apiError{errorExec, err}, nil, nil}
+		}
+
+		matches, err, warnings := selectSeriesByMatchers(q, [][]*labels.Matcher{selector})
```
Comment (krasi-georgiev) [inline web/api/v1/api.go]:
```suggestion
		sets, err, warnings := selectSeriesByMatchers(q, [][]*labels.Matcher{selector})
```

File: web/api/v1/api.go [hunk 3]
```diff
@@ -477,13 +477,93 @@ func (api *API) labelNames(r *http.Request) apiFuncResult {
...
+			return apiFuncResult{nil, &apiError{errorExec, err}, warnings, nil}
+		}
+
+		series := storage.NewMergeSeriesSet(matches, nil)
+		result := make(map[string][]string)
+		for series.Next() {
+			label := series.At().Labels().Get(name)
+			if label != "" && result[label] == nil {
+				result[name] = append(result[name], label)
+				result[label] = result[name] // Prevent from adding same label value twice.
```
Comment (krasi-georgiev) [inline web/api/v1/api.go]: I think we should be able to simplify this a bit, but lets have a look at the tests and will revisit

File: web/api/v1/api.go (hunk 1 above)
Comment (boyskila) [reply web/api/v1/api.go]: Yes, good catch, thanks!

Comment (boyskila) [reply web/api/v1/api.go]: From the other side returning ```Time``` is more flexible - you can take advantage of its methods but I guess opposite conversion is also possible(from ```int64``` to ```Time```)

**Comment by gouthamve** [comment]
@krasi-georgiev Could you take another look at this again?

**Comment by krasi-georgiev** [comment]
That is quite outdated now, I think https://github.com/prometheus/prometheus/pull/7496 takes over this one, but I don't think I will be able to spend any time for a review - working on some Thanos tasks.
